## __5. Data Visualization :__
5.1 Numerical Analysis<br>
5.2 Graphical Analysis

#### __5.1 Numerical Analysis :__
5.1.1 Univariate Analysis <br>
5.1.2 Bivariate Analysis<br>

#### 5.1 Univariate Analysis

1. Univariate analysis involves the <font color = magenta>examination of a single variable in isolation</font>.
2. It allows us to gain insights into the<font color = magenta> distribution, central tendency, and variability of that variable. </font>.
3. It is useful when <font color = magenta>dealing with categorical or discrete data, or when a simplified summary is needed</font>..

1. Value Counts - This technique involves counting the occurence of each               unique
Binning
Correlation:
Skewness:
Kurtosis:
Mean:
|Name|Roll|
|------|-----|

|Technique |Description|
|------------|-------------|
|1. Value Counts|   This technique involves counting the occurrences of each unique category or value within a variable  |
|2. Binning| technique used to convert continuous or numerical data into categorical or discrete intervals (bins or buckets).    |
|3.Correlation|   <ul><li>Correlation is a statistical measure that quantifies the degree of association or relationship between two variables. </li><li>It is typically used for bivariate analysis but can also be relevant in univariate analysis to explore potential relationships that might exist between different subsets or aspects of a single variable. </li><li>The correlation coefficient, often denoted as "r," ranges from -1 to 1. </li><li>A positive value indicates a positive correlation (as one variable increases, the other tends to increase), while a negative value indicates a negative correlation (as one variable increases, the other tends to decrease).</li> <li>A value close to zero suggests little to no linear relationship.</li>  </ul>|
|4. Skewness|  Skewness is a measure of the asymmetry of the probability distribution of a variable.    |
|5. Kurtosis|  Kurtosis is a measure of the "tailedness" of a probability distribution, indicating how much data is in the tails compared to a normal distribution.   |
|6. Mean|he mean, often referred to as the average, is a central measure of location in a dataset. |

In [19]:
import pandas as pd

In [20]:
df = pd.read_csv("AB_NYC_2019.csv")
df.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  19-10-2018   
1  Entire home/apt    225               1                 45  21-05-2019   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  05-07-2019   
4  Entire home/apt     80              10                  9  19-11-2018   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48906 entries, 0 to 48905
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48906 non-null  int64  
 1   name                            48890 non-null  object 
 2   host_id                         48906 non-null  int64  
 3   host_name                       48885 non-null  object 
 4   neighbourhood_group             48906 non-null  object 
 5   neighbourhood                   48906 non-null  object 
 6   latitude                        48906 non-null  float64
 7   longitude                       48906 non-null  float64
 8   room_type                       48906 non-null  object 
 9   price                           48906 non-null  int64  
 10  minimum_nights                  48906 non-null  int64  
 11  number_of_reviews               48906 non-null  int64  
 12  last_review                     

In [22]:
df.describe()

id       host_id      latitude     longitude         price  \
count  4.890600e+04  4.890600e+04  48906.000000  48906.000000  48906.000000   
mean   1.901287e+07  6.760480e+07     40.728952    -73.952175    152.711324   
std    1.098557e+07  7.860866e+07      0.054529      0.046154    240.128713   
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000   
25%    9.464662e+06  7.809567e+06     40.690100    -73.983080     69.000000   
50%    1.967545e+07  3.078463e+07     40.723080    -73.955685    106.000000   
75%    2.915085e+07  1.074344e+08     40.763120    -73.936283    175.000000   
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    48906.000000       48906.000000       38854.000000   
mean         7.031612          23.300454           1.373151   
std         20.512489          44.607175           1.680270   
min          1.000000           0.000000           0.010000   
25%          1.000000           1.000000           0.190000   
50%          3.000000           5.000000           0.720000   
75%          5.000000          24.000000           2.020000   
max       1250.000000         629.000000          58.500000   

       calculated_host_listings_count  availability_365  
count                    48906.000000      48906.000000  
mean                         7.142702        112.782031  
std                         32.948926        131.620370  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000

> Many Colums have dtype lets convert into string

In [25]:
df['host_id']=df['host_id'].astype(str)
df['id']=df['id'].astype(str)
df["latitude"]=df["latitude"].astype(str)
df["longitude"]=df["longitude"].astype(str)

In [26]:
df.describe()

price  minimum_nights  number_of_reviews  reviews_per_month  \
count  48906.000000    48906.000000       48906.000000       38854.000000   
mean     152.711324        7.031612          23.300454           1.373151   
std      240.128713       20.512489          44.607175           1.680270   
min        0.000000        1.000000           0.000000           0.010000   
25%       69.000000        1.000000           1.000000           0.190000   
50%      106.000000        3.000000           5.000000           0.720000   
75%      175.000000        5.000000          24.000000           2.020000   
max    10000.000000     1250.000000         629.000000          58.500000   

       calculated_host_listings_count  availability_365  
count                    48906.000000      48906.000000  
mean                         7.142702        112.782031  
std                         32.948926        131.620370  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000

In [17]:
df.isnull().sum()

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [23]:
10052+38854.000000	

48906.0

<h3 style =" color:#00d5ff;">Let us find the unique values in each column</h3>

In [30]:
df.nunique()


id                                48895
name                              47896
host_id                           37457
host_name                         11452
neighbourhood_group                   5
neighbourhood                       221
latitude                          19048
longitude                         14718
room_type                             3
price                               674
minimum_nights                      109
number_of_reviews                   394
last_review                        1764
reviews_per_month                   937
calculated_host_listings_count       47
availability_365                    366
dtype: int64

id have unique values 48895 but count of id is  48906 so lets check is the gap betwwen observation due to duplicated values

In [33]:
48895+df.duplicated().sum()

np.int64(48906)

In [36]:
df['host_name'].duplicated().sum()+11452

np.int64(48905)

<h4 style =" color:#00d5ff;">Let us find the value count of <b style =" color:magenta" >room type<b></h4>

In [40]:
df['room_type'].value_counts()

room_type
Entire home/apt    25414
Private room       22332
Shared room         1160
Name: count, dtype: int64

In [41]:
df["room_type"].value_counts(normalize = True)

room_type
Entire home/apt    0.519650
Private room       0.456631
Shared room        0.023719
Name: proportion, dtype: float64

<h4 style =" color:#00d5ff;">Let us find the value count of <b style =" color:magenta" >neighbourhood_group<b></h4>

In [42]:
df["neighbourhood_group"].value_counts()

neighbourhood_group
Manhattan        21669
Brooklyn         20107
Queens            5666
Bronx             1091
Staten Island      373
Name: count, dtype: int64

In [44]:
df['price'].value_counts()

price
100     2051
150     2048
50      1534
60      1459
200     1402
        ... 
930        1
920        1
4100       1
3512       1
1494       1
Name: count, Length: 674, dtype: int64

<p style =  "color:crimson; font-size:20px"><b>since price have a widhrange of value so we will divide it into bins for better analysis</b></p>

In [57]:
df['price'].value_counts(bins=10)

(-10.001, 1000.0]    48667
(1000.0, 2000.0]       153
(2000.0, 3000.0]        41
(3000.0, 4000.0]        13
(4000.0, 5000.0]        12
(9000.0, 10000.0]        6
(6000.0, 7000.0]         5
(5000.0, 6000.0]         4
(7000.0, 8000.0]         4
(8000.0, 9000.0]         1
Name: count, dtype: int64

<h4 style =" color:#00d5ff;">Let us find the mean ,median of <b  >data<b></h4>

In [62]:
print(df['price'].mean())
print(df['price'].median())

152.71132376395533
106.0


<h4 style =" color:#00d5ff;">Let us find the correlation of <b  >data<b></h4>

In [70]:
df.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'last_review',
       'reviews_per_month', 'calculated_host_listings_count',
       'availability_365'],
      dtype='object')

In [76]:
temp= pd.DataFrame()

In [77]:
temp["price"] = df["price"]
temp["minimum_nights"]=df["minimum_nights"]
temp["number_of_reviews	"]=df["number_of_reviews"]
temp["reviews_per_month"]=df["reviews_per_month"]
temp["calculated_host_listings_count"]=df["calculated_host_listings_count"]	
temp["availability_365"] =df["availability_365"]

In [79]:
temp.corr()

price  minimum_nights  number_of_reviews\t  \
price                           1.000000        0.042771            -0.048014   
minimum_nights                  0.042771        1.000000            -0.080093   
number_of_reviews\t            -0.048014       -0.080093             1.000000   
reviews_per_month              -0.030608       -0.121772             0.549291   
calculated_host_listings_count  0.057478        0.127917            -0.072375   
availability_365                0.081817        0.144146             0.172002   

                                reviews_per_month  \
price                                   -0.030608   
minimum_nights                          -0.121772   
number_of_reviews\t                      0.549291   
reviews_per_month                        1.000000   
calculated_host_listings_count          -0.009414   
availability_365                         0.185818   

                                calculated_host_listings_count  \
price                                                 0.057478   
minimum_nights                                        0.127917   
number_of_reviews\t                                  -0.072375   
reviews_per_month                                    -0.009414   
calculated_host_listings_count                        1.000000   
availability_365                                      0.225680   

                                availability_365  
price                                   0.081817  
minimum_nights                          0.144146  
number_of_reviews\t                     0.172002  
reviews_per_month                       0.185818  
calculated_host_listings_count          0.225680  
availability_365                        1.000000

<h4 style =" color:#00d5ff;">Let us find the Measure of Spread of <b  >data<b></h4>

In [80]:
temp.skew()

price                             19.120832
minimum_nights                    21.817416
number_of_reviews\t                3.692683
reviews_per_month                  3.130418
calculated_host_listings_count     7.934099
availability_365                   0.763381
dtype: float64

In [81]:
temp.kurt()

price                             585.793048
minimum_nights                    853.553572
number_of_reviews\t                19.553221
reviews_per_month                  42.499825
calculated_host_listings_count     67.567017
availability_365                   -0.997556
dtype: float64